<a href="https://colab.research.google.com/github/Sumanth718/ML-Project/blob/main/Project_Step1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [31]:
## Imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.datasets import load_iris, load_wine
from sklearn.tree import DecisionTreeClassifier, plot_tree, export_graphviz
from sklearn.model_selection import train_test_split, KFold
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.linear_model import LogisticRegression
import os
print('Libraries loaded')

Libraries loaded


In [32]:
import os

drive_path = '/content/drive/MyDrive/'

print(f"Listing files and directories in '{drive_path}':")
if os.path.exists(drive_path):
    for item in os.listdir(drive_path):
        print(item)
else:
    print(f"Error: The path '{drive_path}' does not exist. Please ensure Google Drive is mounted correctly.")

Listing files and directories in '/content/drive/MyDrive/':
Colab Notebooks
Priject_results
UNSW.csv.csv
IDS_project
outputs
.ipynb_checkpoints
Lab6_results


In [33]:
from google.colab import drive
drive.mount('/content/drive')

# IMPORTANT: Replace 'path/to/yohttps://colab.research.google.com/drive/1sx0oYxwyEGTekZ3FeZEc0uWbU5-82Meu#ur/UNSW.csv' with the actual path to your file in Google Drive.
# For example: '/content/drive/MyDrive/UNSW.csv.csv' or '/content/drive/MyDrive/my_data/UNSW.csv.scv'
unsw_file_path = '/content/drive/MyDrive/UNSW.csv.csv' # <--- CHANGE THIS PATH

import pandas as pd
from IPython.display import display

try:
    df_unsw = pd.read_csv(unsw_file_path)
    print('UNSW.csv loaded into df_unsw successfully from Google Drive.')
    print('DataFrame shape:', df_unsw.shape)
    print('First 5 rows of df_unsw:')
    display(df_unsw.head())
except FileNotFoundError:
    print(f"Error: '{unsw_file_path}' not found. Please ensure the file is uploaded to your Google Drive and the path is correct.")
    df_unsw = pd.DataFrame() # Create an empty DataFrame if file not found
    print("An empty DataFrame 'df_unsw' has been created.")
except Exception as e:
    print(f"An error occurred while loading the file: {e}")
    df_unsw = pd.DataFrame() # Create an empty DataFrame in case of other errors
    print("An empty DataFrame 'df_unsw' has been created due to an error.")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


/tmp/ipython-input-1487721004.py:12: DtypeWarning: Columns (3,39,47) have mixed types. Specify dtype option on import or set low_memory=False.
  df_unsw = pd.read_csv(unsw_file_path)


UNSW.csv loaded into df_unsw successfully from Google Drive.
DataFrame shape: (700000, 49)
First 5 rows of df_unsw:


59.166.0.0   6055  149.171.126.5  54145  tcp  FIN  0.072974  4238  60788  \
0  59.166.0.0   7832  149.171.126.3   5607  tcp  FIN  0.144951  5174  91072   
1  59.166.0.8  11397  149.171.126.6     21  tcp  FIN  0.116107  2934   3742   
2  59.166.0.0   3804  149.171.126.3     53  udp  CON  0.000986   146    178   
3  59.166.0.8  14339  149.171.126.6  14724  tcp  FIN  0.038480  8928    320   
4  59.166.0.8  39094  149.171.126.3     53  udp  CON  0.001026   130    162   

   31  ...  0.6  13  13.1  6  7.1  1  1.1  2  Unnamed: 47  0.7  
0  31  ...    0  13    13  6    7  1    1  2          NaN    0  
1  31  ...    1   1     2  7    5  1    1  4          NaN    0  
2  31  ...    0  13    13  6    7  1    1  2          NaN    0  
3  31  ...    0   8    20  7    5  1    1  4          NaN    0  
4  31  ...    0   8    13  6    5  1    1  1          NaN    0  

[5 rows x 49 columns]

## III. Project — Step 1 (Scaffold)

This section is a scaffold to help you complete the project tasks required by the lab:

1. Descriptive analysis of your data.
2. Implementation of necessary preprocessing.
3. Formalisation of the problem.
4. Selection of a baseline model and implementation.

Fill this section with your dataset and details. Example code is provided.

In [34]:
def descriptive_analysis(df):
    print('Shape:', df.shape)
    display(df.head())
    display(df.describe(include='all'))
    print('\nMissing values:\n', df.isna().sum())

# Example preprocessing
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.pipeline import Pipeline

# Example baseline model building function
from sklearn.linear_model import LogisticRegression

def build_baseline_classifier(X, y):
    # Simple baseline: StandardScaler + LogisticRegression
    pipe = Pipeline([('scaler', StandardScaler()), ('clf', LogisticRegression(max_iter=500))])
    pipe.fit(X, y)
    return pipe

print('Project scaffold ready. Replace with your dataset and details.')

Project scaffold ready. Replace with your dataset and details.


# Task
*   **Perform Descriptive Analysis**:
    The `df_unsw` DataFrame is already loaded.
*   **Formalize Machine Learning Problem**:
    The features (X) and target variable (y) are already identified.
*   **Preprocess Data**:
    The data preprocessing steps, including handling mixed types, re-identifying numerical and categorical columns, and applying the `ColumnTransformer` to create `X_preprocessed`, have been completed.
*   **Build and Evaluate Baseline Model**:
    The Logistic Regression baseline model has been trained and evaluated using the preprocessed data.

Based on the plan to re-execute these steps and then provide a comprehensive summary, I will regenerate the necessary code to display the descriptive analysis, problem formalization, preprocessing details, and baseline model performance. After that, I will provide the consolidated summary.

```python
import pandas as pd
import numpy as np
from IPython.display import display

from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer


# --- 1. Perform Descriptive Analysis ---
print("--- 1. Descriptive Analysis of df_unsw ---")
print('Shape:', df_unsw.shape)
print('First 5 rows of df_unsw:')
display(df_unsw.head())
print('\nDescriptive statistics (all columns):')
display(df_unsw.describe(include='all'))
print('\nMissing values:\n', df_unsw.isna().sum())
print("-" * 50)


# --- 2. Formalize Machine Learning Problem ---
print("\n--- 2. Machine Learning Problem Formalization ---")
print("The identified machine learning problem is a classification task.")

# Identify the target variable (y)
y_column_name = df_unsw.columns[-1] # Assuming the last column is the target as per instruction
y = df_unsw[y_column_name]

# Create features (X) by dropping the target column
X = df_unsw.drop(columns=[y_column_name])

print(f"\nChosen target variable (y): '{y_column_name}'")
print(f"Shape of features (X): {X.shape}")
print(f"Shape of target (y): {y.shape}")

# Display unique values and their counts for y
print("\nUnique values and counts for the target variable (y):")
display(y.value_counts())
print("-" * 50)


# --- 3. Preprocess Data ---
print("\n--- 3. Data Preprocessing ---")

# Handle mixed types by converting potentially numeric object columns to numeric
potentially_numeric_object_cols = ['54145', '0.6', 'Unnamed: 47']
for col in potentially_numeric_object_cols:
    if col in X.columns and X[col].dtype == 'object':
        X[col] = pd.to_numeric(X[col], errors='coerce')

# Re-identify numerical and categorical columns after type conversion
numerical_cols = X.select_dtypes(include=np.number).columns
categorical_cols = X.select_dtypes(include='object').columns

print(f"Updated Numerical columns ({len(numerical_cols)}):\n{list(numerical_cols)}")
print(f"\nUpdated Categorical columns ({len(categorical_cols)}):\n{list(categorical_cols)}")

# Preprocessing for numerical data: impute missing values with mean and then scale
numerical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),
    ('scaler', StandardScaler())
])

# Preprocessing for categorical data: impute missing values with most frequent and then one-hot encode
categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

# Create a column transformer to apply different transformations to different columns
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numerical_transformer, numerical_cols),
        ('cat', categorical_transformer, categorical_cols)
    ])

# Create a preprocessing pipeline
preprocessing_pipeline = Pipeline(steps=[('preprocessor', preprocessor)])

# Now fit and transform the data
X_preprocessed = preprocessing_pipeline.fit_transform(X)

print(f"\nOriginal features shape: {X.shape}")
print(f"Preprocessed features shape: {X_preprocessed.shape}")
print("Data preprocessing complete. X_preprocessed contains the transformed features.")
print("-" * 50)


# --- 4. Build and Evaluate Baseline Model ---
print("\n--- 4. Baseline Model (Logistic Regression) Evaluation ---")

# Define the baseline model building function
def build_baseline_classifier(X_transformed, y_target):
    model = LogisticRegression(max_iter=1000, random_state=42, solver='lbfgs', multi_class='auto')
    model.fit(X_transformed, y_target)
    return model

# Split the preprocessed data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_preprocessed, y, test_size=0.3, random_state=42, stratify=y)

print(f"Training set shape: {X_train.shape}, {y_train.shape}")
print(f"Testing set shape: {X_test.shape}, {y_test.shape}")

# Build and train the baseline classifier
baseline_model = build_baseline_classifier(X_train, y_train)

# Make predictions on the test set
y_pred = baseline_model.predict(X_test)

# Evaluate the baseline model
print('\nBaseline Model (Logistic Regression) Evaluation:')
print(f'Accuracy: {accuracy_score(y_test, y_pred):.4f}')
print('\nClassification Report:')
print(classification_report(y_test, y_pred))
print("Baseline model implementation and evaluation complete.")
print("-" * 50)
```
---
## Comprehensive Summary of Initial Findings, Problem Formalization, and Baseline Model Performance

This document summarizes the initial data analysis, formalization of the machine learning problem, and the performance of a baseline model on the `UNSW.csv` dataset.

### 1. Data Analysis Key Findings

*   **Dataset Shape**: The `df_unsw` dataset contains `700,000` rows and `49` columns.
*   **Head of DataFrame**:
    ```
                59.166.0.0   6055  149.171.126.5  54145  tcp  ...   1  1.1   2  Unnamed: 47   0.7
    0           59.166.0.0   7832  149.171.126.3   5607  tcp  ...   1    1   2          NaN     0
    1           59.166.0.8  11397  149.171.126.6     21  tcp  ...   1    1   4          NaN     0
    2           59.166.0.0   3804  149.171.126.3     53  udp  ...   1    1   2          NaN     0
    3           59.166.0.8  14339  149.171.126.6  14724  tcp  ...   1    1   4          NaN     0
    4           59.166.0.8  39094  149.171.126.3     53  udp  ...   1    1   1          NaN     0
    ```
*   **Missing Values**: Significant missing values were identified in three columns:
    *   `0.4`: `285,256` missing values
    *   `0.5`: `307,038` missing values
    *   `Unnamed: 47`: `647,251` missing values (approximately 92% of the column)

### 2. Problem Formalization

*   **Problem Type**: The machine learning problem is formalized as a **binary classification task**.
*   **Target Variable (y)**: The last column, named `'0.7'`, was chosen as the target variable.
    *   Shape of target (y): `(700,000,)`
*   **Features (X)**: All other `48` columns were designated as features.
    *   Shape of features (X): `(700,000, 48)`
*   **Target Variable Distribution**: The target variable exhibits significant class imbalance:
    *   Class `0`: `647,251` instances
    *   Class `1`: `52,749` instances

### 3. Data Preprocessing

*   **Mixed Type Handling**: Columns like `'54145'`, `'0.6'`, and `'Unnamed: 47'` were identified as potentially numeric but loaded as object types due to mixed data. These columns were explicitly converted to numeric, with errors coerced to `NaN`.
*   **Column Re-identification**: After type conversion, the dataset was identified to have `43` numerical columns and `5` categorical columns.
    *   Numerical columns: `['6055', '54145', '0.072974', '4238', '60788', '31', '29', '7', '30', '458245.4375', '6571546.5', '72', '72.1', '255', '255.1', '1003293149', '1003585034', '59', '844', '0', '0.1', '62.04531', '61.899776', '1421955842', '1421955842.1', '1.02269', '0.997042', '0.002317', '0.002173', '0.000144', '0.2', '0.3', '0.4', '0.5', '0.6', '13', '13.1', '6', '7.1', '1', '1.1', '2', 'Unnamed: 47']`
    *   Categorical columns: `['59.166.0.0', '149.171.126.5', 'tcp', 'FIN', '-']`
*   **Preprocessing Pipeline**: A `ColumnTransformer` was used within a `Pipeline`:
    *   **Numerical Features**: Missing values were imputed with the mean, followed by `StandardScaler`.
    *   **Categorical Features**: Missing values were imputed with the most frequent value, followed by `OneHotEncoder`.
*   **Preprocessed Data Shape**: The original `(700,000, 48)` feature DataFrame `X` was transformed into `X_preprocessed` with shape `(700,000, 289)`.

### 4. Baseline Model Performance

*   **Model**: Logistic Regression (`max_iter=1000`, `solver='lbfgs'`).
*   **Data Split**: The preprocessed data was split into training (`70%`) and testing (`30%`) sets, with stratification to maintain class distribution.
    *   Training set shape: `X_train: (490000, 289)`, `y_train: (490000,)`
    *   Testing set shape: `X_test: (210000, 289)`, `y_test: (210000,)`
*   **Evaluation Metrics**:
    *   **Accuracy**: `0.9928`
    *   **Classification Report**:
        ```
                      precision    recall  f1-score   support

               0       1.00      1.00      1.00    194175
               1       0.94      0.96      0.95     15825

            accuracy                           0.99    210000
           macro avg       0.97      0.98      0.97    210000
        weighted avg       0.99      0.99      0.99    210000
        ```
*   **Observations**:
    *   The baseline Logistic Regression model achieved a very high overall accuracy of `0.9928`.
    *   Performance for class `0` (majority class) is almost perfect across all metrics (precision, recall, f1-score of `1.00`).
    *   Performance for class `1` (minority class) is also strong, with a precision of `0.94`, recall of `0.96`, and f1-score of `0.95`. This indicates that despite the class imbalance, the model is doing a reasonably good job of identifying the minority class.

### Next Steps / Considerations

*   **Class Imbalance**: While the current baseline performs well, the significant class imbalance should be kept in mind for more complex models, as it can sometimes lead to models that prioritize the majority class. Techniques like oversampling (SMOTE), undersampling, or using algorithms inherently robust to imbalance could be explored.
*   **Feature Importance**: Investigating feature importance for the Logistic Regression model could provide insights into which features are most influential in predicting the target variable.
*   **Further Model Exploration**: While the baseline is strong, exploring other classification algorithms (e.g., Decision Trees, Random Forests, Gradient Boosting) could lead to further performance improvements or provide a more robust model.
*   **Hyperparameter Tuning**: Optimize the hyperparameters of the Logistic Regression model or other chosen models.

## Perform Descriptive Analysis

### Subtask:
Re-execute the descriptive analysis of the `df_unsw` DataFrame, including shape, head, descriptive statistics, and missing values count.


**Reasoning**:
The subtask explicitly asks to re-execute the descriptive analysis of the `df_unsw` DataFrame, which includes printing its shape, displaying the head, descriptive statistics, and missing value counts. The `descriptive_analysis` function previously defined in the notebook (and present in the current kernel state in cell `035f81e5`) performs all these actions. Therefore, calling this function on the `df_unsw` DataFrame will fulfill all requirements of the subtask.



In [35]:
import pandas as pd
from IPython.display import display

def descriptive_analysis(df):
    print('Shape:', df.shape)
    if df.empty:
        print("DataFrame is empty, skipping detailed description and missing values check.")
        display(df.head())
    else:
        display(df.head())
        display(df.describe(include='all'))
        print('\nMissing values:\n', df.isna().sum())

# Re-execute descriptive analysis on the df_unsw
descriptive_analysis(df_unsw)

Shape: (700000, 49)


59.166.0.0   6055  149.171.126.5  54145  tcp  FIN  0.072974  4238  60788  \
0  59.166.0.0   7832  149.171.126.3   5607  tcp  FIN  0.144951  5174  91072   
1  59.166.0.8  11397  149.171.126.6     21  tcp  FIN  0.116107  2934   3742   
2  59.166.0.0   3804  149.171.126.3     53  udp  CON  0.000986   146    178   
3  59.166.0.8  14339  149.171.126.6  14724  tcp  FIN  0.038480  8928    320   
4  59.166.0.8  39094  149.171.126.3     53  udp  CON  0.001026   130    162   

   31  ...  0.6  13  13.1  6  7.1  1  1.1  2  Unnamed: 47  0.7  
0  31  ...    0  13    13  6    7  1    1  2          NaN    0  
1  31  ...    1   1     2  7    5  1    1  4          NaN    0  
2  31  ...    0  13    13  6    7  1    1  2          NaN    0  
3  31  ...    0   8    20  7    5  1    1  4          NaN    0  
4  31  ...    0   8    13  6    5  1    1  1          NaN    0  

[5 rows x 49 columns]

59.166.0.0           6055  149.171.126.5   54145     tcp     FIN  \
count       700000  700000.000000         700000  700000  700000  700000   
unique          40            NaN             45  101736     134      15   
top     59.166.0.0            NaN  149.171.126.2      53     tcp     FIN   
freq         61811            NaN          61663   99207  455085  452052   
mean           NaN   31330.022319            NaN     NaN     NaN     NaN   
std            NaN   19896.179017            NaN     NaN     NaN     NaN   
min            NaN       0.000000            NaN     NaN     NaN     NaN   
25%            NaN   13498.000000            NaN     NaN     NaN     NaN   
50%            NaN   31875.500000            NaN     NaN     NaN     NaN   
75%            NaN   47455.000000            NaN     NaN     NaN     NaN   
max            NaN   65535.000000            NaN     NaN     NaN     NaN   

             0.072974          4238         60788             31  ...  \
count   700000.000000  7.000000e+05  7.000000e+05  700000.000000  ...   
unique            NaN           NaN           NaN            NaN  ...   
top               NaN           NaN           NaN            NaN  ...   
freq              NaN           NaN           NaN            NaN  ...   
mean         0.583991  4.439205e+03  3.821118e+04      49.803309  ...   
std          3.117189  5.052684e+04  1.595661e+05      60.161503  ...   
min          0.000000  0.000000e+00  0.000000e+00       0.000000  ...   
25%          0.001120  2.640000e+02  1.780000e+02      31.000000  ...   
50%          0.020339  1.580000e+03  2.456000e+03      31.000000  ...   
75%          0.226749  3.598000e+03  1.629000e+04      31.000000  ...   
max         59.999989  1.435577e+07  1.465753e+07     255.000000  ...   

             0.6             13           13.1              6            7.1  \
count   700000.0  700000.000000  700000.000000  700000.000000  700000.000000   
unique       9.0            NaN            NaN            NaN            NaN   
top          0.0            NaN            NaN            NaN            NaN   
freq    370147.0            NaN            NaN            NaN            NaN   
mean         NaN       6.727760       6.549250       4.513316       5.035917   
std          NaN       7.483271       7.438402       5.309101       5.489196   
min          NaN       1.000000       1.000000       1.000000       1.000000   
25%          NaN       2.000000       2.000000       2.000000       2.000000   
50%          NaN       4.000000       4.000000       3.000000       4.000000   
75%          NaN       8.000000       8.000000       5.000000       6.000000   
max          NaN      63.000000      62.000000      60.000000      60.000000   

                    1            1.1              2  Unnamed: 47  \
count   700000.000000  700000.000000  700000.000000        52749   
unique            NaN            NaN            NaN            9   
top               NaN            NaN            NaN      Generic   
freq              NaN            NaN            NaN        27883   
mean         2.609281       1.928206       4.062483          NaN   
std          5.407286       3.862126       7.393228          NaN   
min          1.000000       1.000000       1.000000          NaN   
25%          1.000000       1.000000       1.000000          NaN   
50%          1.000000       1.000000       2.000000          NaN   
75%          1.000000       1.000000       3.000000          NaN   
max         60.000000      60.000000      63.000000          NaN   

                  0.7  
count   700000.000000  
unique            NaN  
top               NaN  
freq              NaN  
mean         0.075356  
std          0.263965  
min          0.000000  
25%          0.000000  
50%          0.000000  
75%          0.000000  
max          1.000000  

[11 rows x 49 columns]


Missing values:
 59.166.0.0            0
6055                  0
149.171.126.5         0
54145                 0
tcp                   0
FIN                   0
0.072974              0
4238                  0
60788                 0
31                    0
29                    0
7                     0
30                    0
-                     0
458245.4375           0
6571546.5             0
72                    0
72.1                  0
255                   0
255.1                 0
1003293149            0
1003585034            0
59                    0
844                   0
0                     0
0.1                   0
62.04531              0
61.899776             0
1421955842            0
1421955842.1          0
1.02269               0
0.997042              0
0.002317              0
0.002173              0
0.000144              0
0.2                   0
0.3                   0
0.4              285256
0.5              307038
0.6                   0
13                    

## Formalize Machine Learning Problem

### Subtask:
Re-execute the identification of features (X) and target variable (y) and display the target variable's class distribution.


**Reasoning**:
The subtask requires re-identifying the features (X) and target variable (y) and displaying the target variable's class distribution. I will regenerate the code to perform these steps, as requested in the instructions.



In [36]:
print("The identified machine learning problem is a classification task.")

# Identify the target variable (y)
y_column_name = df_unsw.columns[-1] # Assuming the last column is the target as per instruction
y = df_unsw[y_column_name]

# Create features (X) by dropping the target column
X = df_unsw.drop(columns=[y_column_name])

print(f"\nChosen target variable (y): '{y_column_name}'")
print(f"Shape of features (X): {X.shape}")
print(f"Shape of target (y): {y.shape}")

# Display unique values and their counts for y
print("\nUnique values and counts for the target variable (y):")
display(y.value_counts())

The identified machine learning problem is a classification task.

Chosen target variable (y): '0.7'
Shape of features (X): (700000, 48)
Shape of target (y): (700000,)

Unique values and counts for the target variable (y):


0.7
0    647251
1     52749
Name: count, dtype: int64

## Preprocess Data

### Subtask:
Re-execute the data preprocessing steps, including handling mixed types, re-identifying numerical and categorical columns, and applying the ColumnTransformer to create X_preprocessed.


**Reasoning**:
The subtask requires re-executing the data preprocessing steps, including handling mixed types, re-identifying column types, and applying a ColumnTransformer. The code will perform these actions sequentially, as specified in the instructions, to create the X_preprocessed DataFrame.



In [37]:
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer

# --- 3. Data Preprocessing ---
print("--- 3. Data Preprocessing ---")

# 1. Handle mixed types by converting potentially numeric object columns to numeric
potentially_numeric_object_cols = ['54145', '0.6', 'Unnamed: 47'] # These were identified in previous run
for col in potentially_numeric_object_cols:
    if col in X.columns and X[col].dtype == 'object':
        X[col] = pd.to_numeric(X[col], errors='coerce')

# 2. Re-identify numerical and categorical columns after type conversion
numerical_cols = X.select_dtypes(include=np.number).columns
categorical_cols = X.select_dtypes(include='object').columns

print(f"Updated Numerical columns ({len(numerical_cols)}):\n{list(numerical_cols)}")
print(f"\nUpdated Categorical columns ({len(categorical_cols)}):\n{list(categorical_cols)}")

# 3. Preprocessing for numerical data: impute missing values with mean and then scale
numerical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),
    ('scaler', StandardScaler())
])

# 4. Preprocessing for categorical data: impute missing values with most frequent and then one-hot encode
categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

# 5. Create a column transformer to apply different transformations to different columns
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numerical_transformer, numerical_cols),
        ('cat', categorical_transformer, categorical_cols)
    ])

# 6. Apply the ColumnTransformer to the features DataFrame X to create X_preprocessed
X_preprocessed = preprocessor.fit_transform(X)

# 7. Print the original features shape and the preprocessed features shape
print(f"\nOriginal features shape: {X.shape}")
print(f"Preprocessed features shape: {X_preprocessed.shape}")
print("Data preprocessing complete. X_preprocessed contains the transformed features.")
print("-" * 50)

--- 3. Data Preprocessing ---
Updated Numerical columns (43):
['6055', '54145', '0.072974', '4238', '60788', '31', '29', '7', '30', '458245.4375', '6571546.5', '72', '72.1', '255', '255.1', '1003293149', '1003585034', '59', '844', '0', '0.1', '62.04531', '61.899776', '1421955842', '1421955842.1', '1.02269', '0.997042', '0.002317', '0.002173', '0.000144', '0.2', '0.3', '0.4', '0.5', '0.6', '13', '13.1', '6', '7.1', '1', '1.1', '2', 'Unnamed: 47']

Updated Categorical columns (5):
['59.166.0.0', '149.171.126.5', 'tcp', 'FIN', '-']


/usr/local/lib/python3.12/dist-packages/sklearn/impute/_base.py:635: UserWarning: Skipping features without any observed values: ['Unnamed: 47']. At least one non-missing value is needed for imputation with strategy='mean'.
  warnings.warn(



Original features shape: (700000, 48)
Preprocessed features shape: (700000, 289)
Data preprocessing complete. X_preprocessed contains the transformed features.
--------------------------------------------------


**Reasoning**:
The previous preprocessing step produced a UserWarning about 'Unnamed: 47' being skipped by SimpleImputer (mean strategy) due to having no observed values. To fix this warning and handle the highly sparse 'Unnamed: 47' column robustly, it will be processed separately with a constant imputation strategy (filling NaNs with 0) before scaling, while other numerical columns will retain their mean imputation.



In [38]:
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer

# --- 3. Data Preprocessing ---
print("--- 3. Data Preprocessing ---")

# 1. Handle mixed types by converting potentially numeric object columns to numeric
potentially_numeric_object_cols = ['54145', '0.6', 'Unnamed: 47']
for col in potentially_numeric_object_cols:
    if col in X.columns and X[col].dtype == 'object':
        X[col] = pd.to_numeric(X[col], errors='coerce')

# 2. Re-identify numerical and categorical columns after type conversion
all_numerical_cols = X.select_dtypes(include=np.number).columns
categorical_cols = X.select_dtypes(include='object').columns

# Separate 'Unnamed: 47' for specific imputation
numerical_cols_for_mean_imputation = [col for col in all_numerical_cols if col != 'Unnamed: 47']
unnamed_47_col = ['Unnamed: 47'] if 'Unnamed: 47' in all_numerical_cols else []

print(f"Numerical columns for mean imputation ({len(numerical_cols_for_mean_imputation)}):\n{list(numerical_cols_for_mean_imputation)}")
print(f"Specific numerical columns for constant imputation ({len(unnamed_47_col)}):\n{list(unnamed_47_col)}")
print(f"\nCategorical columns ({len(categorical_cols)}):\n{list(categorical_cols)}")

transformers = []

# Preprocessing for numerical data (mean imputation and scaling)
if len(numerical_cols_for_mean_imputation) > 0:
    numerical_transformer_mean = Pipeline(steps=[
        ('imputer', SimpleImputer(strategy='mean')),
        ('scaler', StandardScaler())
    ])
    transformers.append(('num_mean', numerical_transformer_mean, numerical_cols_for_mean_imputation))

# Preprocessing for 'Unnamed: 47' (constant imputation and scaling)
if len(unnamed_47_col) > 0:
    numerical_transformer_constant = Pipeline(steps=[
        ('imputer', SimpleImputer(strategy='constant', fill_value=0)), # Address the warning by using constant imputation
        ('scaler', StandardScaler())
    ])
    transformers.append(('num_constant', numerical_transformer_constant, unnamed_47_col))

# Preprocessing for categorical data: impute missing values with most frequent and then one-hot encode
if len(categorical_cols) > 0:
    categorical_transformer = Pipeline(steps=[
        ('imputer', SimpleImputer(strategy='most_frequent')),
        ('onehot', OneHotEncoder(handle_unknown='ignore'))
    ])
    transformers.append(('cat', categorical_transformer, categorical_cols))

# Create a column transformer to apply different transformations to different columns
# Default remainder is 'drop', which is fine as all relevant columns are handled.
preprocessor = ColumnTransformer(
    transformers=transformers
)

# Apply the ColumnTransformer to the features DataFrame X to create X_preprocessed
X_preprocessed = preprocessor.fit_transform(X)

# Print the original features shape and the preprocessed features shape
print(f"\nOriginal features shape: {X.shape}")
print(f"Preprocessed features shape: {X_preprocessed.shape}")
print("Data preprocessing complete. X_preprocessed contains the transformed features.")
print("-" * 50)

--- 3. Data Preprocessing ---
Numerical columns for mean imputation (42):
['6055', '54145', '0.072974', '4238', '60788', '31', '29', '7', '30', '458245.4375', '6571546.5', '72', '72.1', '255', '255.1', '1003293149', '1003585034', '59', '844', '0', '0.1', '62.04531', '61.899776', '1421955842', '1421955842.1', '1.02269', '0.997042', '0.002317', '0.002173', '0.000144', '0.2', '0.3', '0.4', '0.5', '0.6', '13', '13.1', '6', '7.1', '1', '1.1', '2']
Specific numerical columns for constant imputation (1):
['Unnamed: 47']

Categorical columns (5):
['59.166.0.0', '149.171.126.5', 'tcp', 'FIN', '-']


/usr/local/lib/python3.12/dist-packages/sklearn/impute/_base.py:572: FutureWarning: Currently, when `keep_empty_feature=False` and `strategy="constant"`, empty features are not dropped. This behaviour will change in version 1.8. Set `keep_empty_feature=True` to preserve this behaviour.
  warnings.warn(



Original features shape: (700000, 48)
Preprocessed features shape: (700000, 290)
Data preprocessing complete. X_preprocessed contains the transformed features.
--------------------------------------------------


**Reasoning**:
The previous step raised a FutureWarning related to SimpleImputer with `strategy='constant'` and `keep_empty_feature=False`. To suppress this warning and ensure consistent behavior with future scikit-learn versions, the `SimpleImputer` for `unnamed_47_col` will be explicitly updated to set `keep_empty_feature=True`.



In [40]:
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer

# --- 3. Data Preprocessing ---
print("--- 3. Data Preprocessing ---")

# 1. Handle mixed types by converting potentially numeric object columns to numeric
potentially_numeric_object_cols = ['54145', '0.6', 'Unnamed: 47']
for col in potentially_numeric_object_cols:
    if col in X.columns and X[col].dtype == 'object':
        X[col] = pd.to_numeric(X[col], errors='coerce')

# 2. Re-identify numerical and categorical columns after type conversion
all_numerical_cols = X.select_dtypes(include=np.number).columns
categorical_cols = X.select_dtypes(include='object').columns

# Separate 'Unnamed: 47' for specific imputation
numerical_cols_for_mean_imputation = [col for col in all_numerical_cols if col != 'Unnamed: 47']
unnamed_47_col = ['Unnamed: 47'] if 'Unnamed: 47' in all_numerical_cols else []

print(f"Numerical columns for mean imputation ({len(numerical_cols_for_mean_imputation)}):\n{list(numerical_cols_for_mean_imputation)}")
print(f"Specific numerical columns for constant imputation ({len(unnamed_47_col)}):\n{list(unnamed_47_col)}")
print(f"\nCategorical columns ({len(categorical_cols)}):\n{list(categorical_cols)}")

transformers = []

# Preprocessing for numerical data (mean imputation and scaling)
if len(numerical_cols_for_mean_imputation) > 0:
    numerical_transformer_mean = Pipeline(steps=[
        ('imputer', SimpleImputer(strategy='mean')),
        ('scaler', StandardScaler())
    ])
    transformers.append(('num_mean', numerical_transformer_mean, numerical_cols_for_mean_imputation))

# Preprocessing for 'Unnamed: 47' (constant imputation and scaling)
if len(unnamed_47_col) > 0:
    numerical_transformer_constant = Pipeline(steps=[
        ('imputer', SimpleImputer(strategy='constant', fill_value=0)), # Removed keep_empty_feature=True
        ('scaler', StandardScaler())
    ])
    transformers.append(('num_constant', numerical_transformer_constant, unnamed_47_col))

# Preprocessing for categorical data: impute missing values with most frequent and then one-hot encode
if len(categorical_cols) > 0:
    categorical_transformer = Pipeline(steps=[
        ('imputer', SimpleImputer(strategy='most_frequent')),
        ('onehot', OneHotEncoder(handle_unknown='ignore'))
    ])
    transformers.append(('cat', categorical_transformer, categorical_cols))

# Create a column transformer to apply different transformations to different columns
# Default remainder is 'drop', which is fine as all relevant columns are handled.
preprocessor = ColumnTransformer(
    transformers=transformers
)

# Apply the ColumnTransformer to the features DataFrame X to create X_preprocessed
X_preprocessed = preprocessor.fit_transform(X)

# Print the original features shape and the preprocessed features shape
print(f"\nOriginal features shape: {X.shape}")
print(f"Preprocessed features shape: {X_preprocessed.shape}")
print("Data preprocessing complete. X_preprocessed contains the transformed features.")
print("-" * 50)


--- 3. Data Preprocessing ---
Numerical columns for mean imputation (42):
['6055', '54145', '0.072974', '4238', '60788', '31', '29', '7', '30', '458245.4375', '6571546.5', '72', '72.1', '255', '255.1', '1003293149', '1003585034', '59', '844', '0', '0.1', '62.04531', '61.899776', '1421955842', '1421955842.1', '1.02269', '0.997042', '0.002317', '0.002173', '0.000144', '0.2', '0.3', '0.4', '0.5', '0.6', '13', '13.1', '6', '7.1', '1', '1.1', '2']
Specific numerical columns for constant imputation (1):
['Unnamed: 47']

Categorical columns (5):
['59.166.0.0', '149.171.126.5', 'tcp', 'FIN', '-']


/usr/local/lib/python3.12/dist-packages/sklearn/impute/_base.py:572: FutureWarning: Currently, when `keep_empty_feature=False` and `strategy="constant"`, empty features are not dropped. This behaviour will change in version 1.8. Set `keep_empty_feature=True` to preserve this behaviour.
  warnings.warn(



Original features shape: (700000, 48)
Preprocessed features shape: (700000, 290)
Data preprocessing complete. X_preprocessed contains the transformed features.
--------------------------------------------------


**Reasoning**:
The `TypeError` indicates that the `keep_empty_feature` argument is not supported by the version of `SimpleImputer` in this environment. I need to remove this argument to resolve the error and allow the preprocessing to proceed.



In [ ]:
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer

# --- 3. Data Preprocessing ---
print("--- 3. Data Preprocessing ---")

# 1. Handle mixed types by converting potentially numeric object columns to numeric
potentially_numeric_object_cols = ['54145', '0.6', 'Unnamed: 47']
for col in potentially_numeric_object_cols:
    if col in X.columns and X[col].dtype == 'object':
        X[col] = pd.to_numeric(X[col], errors='coerce')

# 2. Re-identify numerical and categorical columns after type conversion
all_numerical_cols = X.select_dtypes(include=np.number).columns
categorical_cols = X.select_dtypes(include='object').columns

# Separate 'Unnamed: 47' for specific imputation
numerical_cols_for_mean_imputation = [col for col in all_numerical_cols if col != 'Unnamed: 47']
unnamed_47_col = ['Unnamed: 47'] if 'Unnamed: 47' in all_numerical_cols else []

print(f"Numerical columns for mean imputation ({len(numerical_cols_for_mean_imputation)}):\n{list(numerical_cols_for_mean_imputation)}")
print(f"Specific numerical columns for constant imputation ({len(unnamed_47_col)}):\n{list(unnamed_47_col)}")
print(f"\nCategorical columns ({len(categorical_cols)}):\n{list(categorical_cols)}")

transformers = []

# Preprocessing for numerical data (mean imputation and scaling)
if len(numerical_cols_for_mean_imputation) > 0:
    numerical_transformer_mean = Pipeline(steps=[
        ('imputer', SimpleImputer(strategy='mean')),
        ('scaler', StandardScaler())
    ])
    transformers.append(('num_mean', numerical_transformer_mean, numerical_cols_for_mean_imputation))

# Preprocessing for 'Unnamed: 47' (constant imputation and scaling)
if len(unnamed_47_col) > 0:
    numerical_transformer_constant = Pipeline(steps=[
        ('imputer', SimpleImputer(strategy='constant', fill_value=0)), # Removed keep_empty_feature=True
        ('scaler', StandardScaler())
    ])
    transformers.append(('num_constant', numerical_transformer_constant, unnamed_47_col))

# Preprocessing for categorical data: impute missing values with most frequent and then one-hot encode
if len(categorical_cols) > 0:
    categorical_transformer = Pipeline(steps=[
        ('imputer', SimpleImputer(strategy='most_frequent')),
        ('onehot', OneHotEncoder(handle_unknown='ignore'))
    ])
    transformers.append(('cat', categorical_transformer, categorical_cols))

# Create a column transformer to apply different transformations to different columns
# Default remainder is 'drop', which is fine as all relevant columns are handled.
preprocessor = ColumnTransformer(
    transformers=transformers
)

# Apply the ColumnTransformer to the features DataFrame X to create X_preprocessed
X_preprocessed = preprocessor.fit_transform(X)

# Print the original features shape and the preprocessed features shape
print(f"\nOriginal features shape: {X.shape}")
print(f"Preprocessed features shape: {X_preprocessed.shape}")
print("Data preprocessing complete. X_preprocessed contains the transformed features.")
print("-" * 50)

**Reasoning**:
The previous code already correctly removes the unsupported `keep_empty_feature` argument from `SimpleImputer`, which resolved the `TypeError`. This code successfully processes the data, and while it produces a `FutureWarning`, it does not block execution or indicate incorrect behavior for the current scikit-learn version.



In [41]:
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer

# --- 3. Data Preprocessing ---
print("--- 3. Data Preprocessing ---")

# 1. Handle mixed types by converting potentially numeric object columns to numeric
potentially_numeric_object_cols = ['54145', '0.6', 'Unnamed: 47']
for col in potentially_numeric_object_cols:
    if col in X.columns and X[col].dtype == 'object':
        X[col] = pd.to_numeric(X[col], errors='coerce')

# 2. Re-identify numerical and categorical columns after type conversion
all_numerical_cols = X.select_dtypes(include=np.number).columns
categorical_cols = X.select_dtypes(include='object').columns

# Separate 'Unnamed: 47' for specific imputation
numerical_cols_for_mean_imputation = [col for col in all_numerical_cols if col != 'Unnamed: 47']
unnamed_47_col = ['Unnamed: 47'] if 'Unnamed: 47' in all_numerical_cols else []

print(f"Numerical columns for mean imputation ({len(numerical_cols_for_mean_imputation)}):\n{list(numerical_cols_for_mean_imputation)}")
print(f"Specific numerical columns for constant imputation ({len(unnamed_47_col)}):\n{list(unnamed_47_col)}")
print(f"\nCategorical columns ({len(categorical_cols)}):\n{list(categorical_cols)}")

transformers = []

# Preprocessing for numerical data (mean imputation and scaling)
if len(numerical_cols_for_mean_imputation) > 0:
    numerical_transformer_mean = Pipeline(steps=[
        ('imputer', SimpleImputer(strategy='mean')),
        ('scaler', StandardScaler())
    ])
    transformers.append(('num_mean', numerical_transformer_mean, numerical_cols_for_mean_imputation))

# Preprocessing for 'Unnamed: 47' (constant imputation and scaling)
if len(unnamed_47_col) > 0:
    numerical_transformer_constant = Pipeline(steps=[
        ('imputer', SimpleImputer(strategy='constant', fill_value=0)),
        ('scaler', StandardScaler())
    ])
    transformers.append(('num_constant', numerical_transformer_constant, unnamed_47_col))

# Preprocessing for categorical data: impute missing values with most frequent and then one-hot encode
if len(categorical_cols) > 0:
    categorical_transformer = Pipeline(steps=[
        ('imputer', SimpleImputer(strategy='most_frequent')),
        ('onehot', OneHotEncoder(handle_unknown='ignore'))
    ])
    transformers.append(('cat', categorical_transformer, categorical_cols))

# Create a column transformer to apply different transformations to different columns
# Default remainder is 'drop', which is fine as all relevant columns are handled.
preprocessor = ColumnTransformer(
    transformers=transformers
)

# Apply the ColumnTransformer to the features DataFrame X to create X_preprocessed
X_preprocessed = preprocessor.fit_transform(X)

# Print the original features shape and the preprocessed features shape
print(f"\nOriginal features shape: {X.shape}")
print(f"Preprocessed features shape: {X_preprocessed.shape}")
print("Data preprocessing complete. X_preprocessed contains the transformed features.")
print("-" * 50)


--- 3. Data Preprocessing ---
Numerical columns for mean imputation (42):
['6055', '54145', '0.072974', '4238', '60788', '31', '29', '7', '30', '458245.4375', '6571546.5', '72', '72.1', '255', '255.1', '1003293149', '1003585034', '59', '844', '0', '0.1', '62.04531', '61.899776', '1421955842', '1421955842.1', '1.02269', '0.997042', '0.002317', '0.002173', '0.000144', '0.2', '0.3', '0.4', '0.5', '0.6', '13', '13.1', '6', '7.1', '1', '1.1', '2']
Specific numerical columns for constant imputation (1):
['Unnamed: 47']

Categorical columns (5):
['59.166.0.0', '149.171.126.5', 'tcp', 'FIN', '-']


/usr/local/lib/python3.12/dist-packages/sklearn/impute/_base.py:572: FutureWarning: Currently, when `keep_empty_feature=False` and `strategy="constant"`, empty features are not dropped. This behaviour will change in version 1.8. Set `keep_empty_feature=True` to preserve this behaviour.
  warnings.warn(



Original features shape: (700000, 48)
Preprocessed features shape: (700000, 290)
Data preprocessing complete. X_preprocessed contains the transformed features.
--------------------------------------------------


**Reasoning**:
The data preprocessing step has been successfully completed, resulting in `X_preprocessed`. The next logical step, as per the overall project plan, is to re-execute the building and evaluation of the baseline model using this preprocessed data. This involves splitting the data into training and testing sets, training a Logistic Regression model, and then evaluating its performance using accuracy and a classification report.



In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, accuracy_score
from sklearn.linear_model import LogisticRegression

# --- 4. Build and Evaluate Baseline Model ---
print("\n--- 4. Baseline Model (Logistic Regression) Evaluation ---")

# Define the baseline model building function
def build_baseline_classifier(X_transformed, y_target):
    model = LogisticRegression(max_iter=1000, random_state=42, solver='lbfgs', multi_class='auto')
    model.fit(X_transformed, y_target)
    return model

# Split the preprocessed data into training and testing sets
# Ensure y is a Series/DataFrame if it's not already, for stratification
if isinstance(y, (pd.Series, pd.DataFrame)):
    X_train, X_test, y_train, y_test = train_test_split(X_preprocessed, y, test_size=0.3, random_state=42, stratify=y)
else:
    # Fallback if y is a numpy array, stratification might need adjustment for multi-class
    X_train, X_test, y_train, y_test = train_test_split(X_preprocessed, y, test_size=0.3, random_state=42)

print(f"Training set shape: {X_train.shape}, {y_train.shape}")
print(f"Testing set shape: {X_test.shape}, {y_test.shape}")

# Build and train the baseline classifier
baseline_model = build_baseline_classifier(X_train, y_train)

# Make predictions on the test set
y_pred = baseline_model.predict(X_test)

# Evaluate the baseline model
print('\nBaseline Model (Logistic Regression) Evaluation:')
print(f'Accuracy: {accuracy_score(y_test, y_pred):.4f}')
print('\nClassification Report:')
print(classification_report(y_test, y_pred))
print("Baseline model implementation and evaluation complete.")
print("-" * 50)


**Reasoning**:
The previous step produced a `FutureWarning` regarding the `multi_class` parameter in `LogisticRegression`, indicating it will be deprecated. To remove this warning and ensure forward compatibility, I will remove the `multi_class='auto'` argument from the `LogisticRegression` constructor, allowing it to use its default behavior.



In [42]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, accuracy_score
from sklearn.linear_model import LogisticRegression

# --- 4. Build and Evaluate Baseline Model ---
print("\n--- 4. Baseline Model (Logistic Regression) Evaluation ---")

# Define the baseline model building function
def build_baseline_classifier(X_transformed, y_target):
    model = LogisticRegression(max_iter=1000, random_state=42, solver='lbfgs') # Removed multi_class='auto'
    model.fit(X_transformed, y_target)
    return model

# Split the preprocessed data into training and testing sets
# Ensure y is a Series/DataFrame if it's not already, for stratification
if isinstance(y, (pd.Series, pd.DataFrame)):
    X_train, X_test, y_train, y_test = train_test_split(X_preprocessed, y, test_size=0.3, random_state=42, stratify=y)
else:
    # Fallback if y is a numpy array, stratification might need adjustment for multi-class
    X_train, X_test, y_train, y_test = train_test_split(X_preprocessed, y, test_size=0.3, random_state=42)

print(f"Training set shape: {X_train.shape}, {y_train.shape}")
print(f"Testing set shape: {X_test.shape}, {y_test.shape}")

# Build and train the baseline classifier
baseline_model = build_baseline_classifier(X_train, y_train)

# Make predictions on the test set
y_pred = baseline_model.predict(X_test)

# Evaluate the baseline model
print('\nBaseline Model (Logistic Regression) Evaluation:')
print(f'Accuracy: {accuracy_score(y_test, y_pred):.4f}')
print('\nClassification Report:')
print(classification_report(y_test, y_pred))
print("Baseline model implementation and evaluation complete.")
print("-" * 50)



--- 4. Baseline Model (Logistic Regression) Evaluation ---
Training set shape: (490000, 290), (490000,)
Testing set shape: (210000, 290), (210000,)

Baseline Model (Logistic Regression) Evaluation:
Accuracy: 0.9929

Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00    194175
           1       0.95      0.96      0.95     15825

    accuracy                           0.99    210000
   macro avg       0.97      0.98      0.97    210000
weighted avg       0.99      0.99      0.99    210000

Baseline model implementation and evaluation complete.
--------------------------------------------------


# Task
## Implement Ensemble Model (RandomForest)

### Subtask:
Initialize and train a RandomForestClassifier. Evaluate its baseline performance using accuracy and a classification report on the test set.

## Implement Ensemble Model (RandomForest)

### Subtask:
Initialize and train a RandomForestClassifier. Evaluate its baseline performance using accuracy and a classification report on the test set.


**Reasoning**:
The subtask requires initializing, training, and evaluating a RandomForestClassifier model. This code block will import the necessary class, create an instance, train it with the preprocessed data, make predictions, and then print the accuracy score and classification report.



In [43]:
from sklearn.ensemble import RandomForestClassifier

print("\n--- Implementing Ensemble Model (RandomForest) ---")

# 1. Initialize a RandomForestClassifier object
# Set random_state for reproducibility and n_estimators as a starting point
rf_model = RandomForestClassifier(n_estimators=100, random_state=42, n_jobs=-1) # n_jobs=-1 to use all available cores

# 2. Train the RandomForestClassifier model
print("Training RandomForestClassifier...")
rf_model.fit(X_train, y_train)
print("RandomForestClassifier training complete.")

# 3. Make predictions on the test features
y_pred_rf = rf_model.predict(X_test)

# 4. Evaluate the model
print('\nRandomForestClassifier Evaluation:')
print(f'Accuracy: {accuracy_score(y_test, y_pred_rf):.4f}')
print('\nClassification Report:')
print(classification_report(y_test, y_pred_rf))
print("--------------------------------------------------")


--- Implementing Ensemble Model (RandomForest) ---
Training RandomForestClassifier...
RandomForestClassifier training complete.

RandomForestClassifier Evaluation:
Accuracy: 0.9973

Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00    194175
           1       0.98      0.98      0.98     15825

    accuracy                           1.00    210000
   macro avg       0.99      0.99      0.99    210000
weighted avg       1.00      1.00      1.00    210000

--------------------------------------------------


## Define Parameter Grid for Grid Search

### Subtask:
Define a dictionary of hyperparameters for the RandomForestClassifier that will be tuned using Grid Search. Include parameters like `n_estimators`, `max_depth`, and `min_samples_leaf`.


**Reasoning**:
The subtask requires defining a dictionary of hyperparameters for the RandomForestClassifier to be used in Grid Search, as specified in the instructions. This dictionary will be named `param_grid` and will include `n_estimators`, `max_depth`, and `min_samples_leaf`.



In [44]:
print("--- Defining Parameter Grid for Grid Search ---")

# Define the parameter grid for RandomForestClassifier
param_grid = {
    'n_estimators': [50, 100, 200],  # Number of trees in the forest
    'max_depth': [10, 20, None],   # Maximum depth of the tree (None means unlimited)
    'min_samples_leaf': [1, 5, 10] # Minimum number of samples required to be at a leaf node
}

print("Parameter grid defined successfully:")
for param, values in param_grid.items():
    print(f"  {param}: {values}")
print("--------------------------------------------------")

--- Defining Parameter Grid for Grid Search ---
Parameter grid defined successfully:
  n_estimators: [50, 100, 200]
  max_depth: [10, 20, None]
  min_samples_leaf: [1, 5, 10]
--------------------------------------------------


## Perform Grid Search on Ensemble Model

### Subtask:
Apply GridSearchCV to the RandomForestClassifier using the defined parameter grid. This step will systematically search for the best combination of hyperparameters that maximize model performance (e.g., accuracy or F1-score).


**Reasoning**:
The subtask requires applying GridSearchCV to the RandomForestClassifier using the previously defined parameter grid. This involves importing `GridSearchCV`, initializing it with the model and parameters, fitting it to the training data, and then printing the best parameters and score found.



In [45]:
from sklearn.model_selection import GridSearchCV

print("--- Applying GridSearchCV to RandomForestClassifier ---")

# 1. Initialize GridSearchCV
# Using cv=3 for quicker execution, and 'accuracy' as the scoring metric.
# n_jobs=-1 to utilize all available cores for parallel processing.
grid_search = GridSearchCV(estimator=rf_model,
                           param_grid=param_grid,
                           cv=3, # Using 3-fold cross-validation
                           scoring='accuracy', # Or 'f1_weighted', 'roc_auc' depending on problem
                           verbose=2, # Increased verbosity to see progress
                           n_jobs=-1) # Use all available cores

# 2. Fit GridSearchCV to the training data
print("Fitting GridSearchCV...")
grid_search.fit(X_train, y_train)
print("GridSearchCV fitting complete.")

# 3. Print the best parameters found
print(f"\nBest parameters found: {grid_search.best_params_}")

# 4. Print the best score obtained
print(f"Best cross-validation accuracy: {grid_search.best_score_:.4f}")
print("--------------------------------------------------")

--- Applying GridSearchCV to RandomForestClassifier ---
Fitting GridSearchCV...
Fitting 3 folds for each of 27 candidates, totalling 81 fits
GridSearchCV fitting complete.

Best parameters found: {'max_depth': None, 'min_samples_leaf': 1, 'n_estimators': 200}
Best cross-validation accuracy: 0.9970
--------------------------------------------------


## Evaluate Best Ensemble Model from Grid Search

### Subtask:
Retrieve the best estimator from the Grid Search results and evaluate its performance on the test set using accuracy and a classification report. Compare these results with the initial baseline ensemble model.


**Reasoning**:
The subtask requires retrieving the best estimator from the completed GridSearchCV, using it to make predictions on the test set, and then evaluating its performance using accuracy and a classification report.



In [46]:
from sklearn.metrics import accuracy_score, classification_report

print("--- Evaluating Best Ensemble Model from Grid Search ---")

# 1. Retrieve the best performing model from GridSearchCV
best_rf_model = grid_search.best_estimator_
print(f"Best RandomForest model found: {best_rf_model}")

# 2. Use the best_rf_model to make predictions on the test set
y_pred_best_rf = best_rf_model.predict(X_test)

# 3. Calculate and print the accuracy score
accuracy_best_rf = accuracy_score(y_test, y_pred_best_rf)
print(f'\nBest RandomForest Model (from Grid Search) Evaluation:')
print(f'Accuracy: {accuracy_best_rf:.4f}')

# 4. Generate and print a classification report
print('\nClassification Report:')
print(classification_report(y_test, y_pred_best_rf))
print("--------------------------------------------------")

--- Evaluating Best Ensemble Model from Grid Search ---
Best RandomForest model found: RandomForestClassifier(n_estimators=200, n_jobs=-1, random_state=42)

Best RandomForest Model (from Grid Search) Evaluation:
Accuracy: 0.9974

Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00    194175
           1       0.98      0.98      0.98     15825

    accuracy                           1.00    210000
   macro avg       0.99      0.99      0.99    210000
weighted avg       1.00      1.00      1.00    210000

--------------------------------------------------


### Comparison with Baseline Ensemble Model

**Initial RandomForestClassifier Performance (Baseline):**
*   **Accuracy**: `0.9973`
*   **Class 0 (Majority)**: Precision: `1.00`, Recall: `1.00`, F1-score: `1.00`
*   **Class 1 (Minority)**: Precision: `0.98`, Recall: `0.98`, F1-score: `0.98`

**Best RandomForestClassifier Performance (from Grid Search):**
*   **Accuracy**: `0.9974`
*   **Best Parameters**: `{'max_depth': None, 'min_samples_leaf': 1, 'n_estimators': 200}`
*   **Class 0 (Majority)**: Precision: `1.00`, Recall: `1.00`, F1-score: `1.00`
*   **Class 1 (Minority)**: Precision: `0.98`, Recall: `0.98`, F1-score: `0.98`

**Observations and Comparison:**

The Grid Search identified `n_estimators=200`, `max_depth=None` (unlimited), and `min_samples_leaf=1` as the best parameters. The initial RandomForest model was created with `n_estimators=100`, `max_depth=None` (default), and `min_samples_leaf=1` (default).

*   **Accuracy**: The accuracy improved slightly from `0.9973` to `0.9974`. While this is a marginal increase, it indicates that tuning the number of estimators (from 100 to 200) had a positive, albeit small, impact.
*   **Precision, Recall, F1-score**: For both classes (0 and 1), the precision, recall, and f1-scores remained excellent and largely unchanged. The model was already performing very well on both classes even before hyperparameter tuning.

**Conclusion:**

The Random Forest model, even with its initial default parameters (or slight tuning), demonstrates very strong performance on this dataset. The Grid Search confirmed that increasing `n_estimators` slightly improved the overall accuracy, but the model was already highly effective. This suggests that the dataset is well-suited for a powerful ensemble method like Random Forest, and further complex tuning might yield diminishing returns given the already near-perfect scores.

## Final Task

### Subtask:
Provide a comprehensive summary of the Grid Search and Ensemble Learning results, including the best hyperparameters found, the performance of the optimized ensemble model, and a comparison with previous baseline models.


## Summary:

### Q&A
The comprehensive summary of the Grid Search and Ensemble Learning results is provided in the "Data Analysis Key Findings" section below.

### Data Analysis Key Findings
*   **Baseline Ensemble Model Performance**: The initial RandomForestClassifier, with `n_estimators=100`, achieved an accuracy of `0.9973` on the test set. Its classification report showed near-perfect scores for both classes (Class 0: Precision, Recall, F1-score all `1.00`; Class 1: Precision, Recall, F1-score all `0.98`).
*   **Hyperparameter Grid for Grid Search**: A parameter grid was defined for tuning the RandomForestClassifier, including `n_estimators` (`[50, 100, 200]`), `max_depth` (`[10, 20, None]`), and `min_samples_leaf` (`[1, 5, 10]`).
*   **Best Hyperparameters from Grid Search**: The GridSearchCV identified the best hyperparameters as `{'max_depth': None, 'min_samples_leaf': 1, 'n_estimators': 200}`. The best cross-validation accuracy obtained during the search was `0.9970`.
*   **Optimized Ensemble Model Performance**: The RandomForestClassifier with the best hyperparameters (`n_estimators=200`, `max_depth=None`, `min_samples_leaf=1`) achieved an accuracy of `0.9974` on the test set. Its classification report maintained excellent performance (Class 0: Precision, Recall, F1-score all `1.00`; Class 1: Precision, Recall, F1-score all `0.98`).
*   **Comparison with Baseline Models**: The optimized ensemble model showed a marginal increase in accuracy from `0.9973` (baseline) to `0.9974`. While the overall metrics (precision, recall, F1-score) remained largely unchanged and exceptionally high for both classes, the Grid Search confirmed that increasing `n_estimators` from 100 to 200 provided a slight performance boost.

### Insights or Next Steps
*   The RandomForest model already demonstrated extremely high performance even with its initial default parameters, suggesting the dataset is highly amenable to this type of ensemble method. Further extensive hyperparameter tuning may yield only marginal gains.
*   Given the near-perfect scores, a next step could involve exploring more challenging datasets or introducing more complex feature engineering to identify if the model's robustness holds up under different conditions, or to confirm if the current dataset is relatively easy to classify.
